src/ProgressMeter.jl

```diff
@@ -57,46 +80,37 @@ the current task. Optionally you can disable the progress bar by setting
 """
 mutable struct Progress <: AbstractProgress
-    n::Int
-    reentrantlocker::Threads.ReentrantLock
-    dt::Float64
-    counter::Int
-    tinit::Float64
-    tsecond::Float64           # ignore the first loop given usually uncharacteristically slow
-    tlast::Float64
-    printed::Bool              # true if we have issued at least one status update
-    desc::String               # prefix to the percentage, e.g.  "Computing..."
+    n::Int                  # total number of iterations
+    start::Int              # which iteration number to start from
     barlen::Union{Int,Nothing} # progress bar size (default is available terminal width)
-    barglyphs::BarGlyphs       # the characters to be used in the bar
-    color::Symbol              # default to green
-    output::IO                 # output stream into which the progress is written
-    offset::Int                # position offset of progress bar (default is 0)
-    numprintedvalues::Int      # num values printed below progress in last iteration
-    start::Int                 # which iteration number to start from
-    enabled::Bool              # is the output enabled
-    showspeed::Bool            # should the output include average time per iteration
-    check_iterations::Int
-    prev_update_count::Int
-    threads_used::Vector{Int}
-
-    function Progress(n::Integer;
-                      dt::Real=0.1,
-                      desc::AbstractString="Progress: ",
-                      color::Symbol=:green,
-                      output::IO=stderr,
-                      barlen=nothing,
-                      barglyphs::BarGlyphs=BarGlyphs('|','█', Sys.iswindows() ? '█' : ['▏','▎','▍','▌','▋','▊','▉'],' ','|',),
-                      offset::Integer=0,
-                      start::Integer=0,
-                      enabled::Bool = true,
-                      showspeed::Bool = false,
-                     )
+    barglyphs::BarGlyphs    # the characters to be used in the bar
+    # internals
+    core::ProgressCore
+
+    function Progress(
+            n::Integer;
+            start::Integer=0,
+            barlen::Union{Int,Nothing}=nothing,
+            barglyphs::BarGlyphs=defaultglyphs,
+            kwargs...)
         CLEAR_IJULIA[] = clear_ijulia()
-        reentrantlocker = Threads.ReentrantLock()
-        counter = start
-        tinit = tsecond = tlast = time()
-        printed = false
-        new(n, reentrantlocker, dt, counter, tinit, tsecond, tlast, printed, desc, barlen, barglyphs, color, output, offset, 0, start, enabled, showspeed, 1, 1, Int[])
+        core = ProgressCore(;kwargs...)
+        new(n, start, barlen, barglyphs, core)
+    end
+end
+# forward common core properties to main types
+function Base.setproperty!(p::T, name::Symbol, value) where T<:AbstractProgress
+    if hasfield(T, name)
+        setfield!(p, name, value)
+    else
+        setproperty!(p.core, name, value)
+    end
+end
```

Comment:
This is a breaking change, because it isn't handling automatic conversion of `value` to `T` like the default `setproperty!` method does: https://github.com/JuliaLang/julia/blob/a69aa30478ac9beda594235ff6f2e39549fdb752/base/Base.jl#L50-L54